### PR TITLE
Verify control master is open for deploy and check_ssh

### PR DESCRIPTION
**Pull request by aaron-collier**
## Why was this change made?

This should help avoid locking accounts out by verifying that a control master is established before running ssh commands.

Prompts the user to allow a pause for connecting a control master, and a `-c, --control_master` option to skip the prompt.

Uploading Screen Recording 2024-12-17 at 3.59.47 PM.mov…

### CODE REVIEW

File: bin/sdr [hunk 1]
```diff
@@ -88,10 +103,17 @@ class CLI < Thor
          banner: 'ENVIRONMENT',
          desc: "Environment (#{Settings.supported_envs.keys.map(&:to_s)})",
          aliases: '-e'
+  option :control_master,
+         type: :boolean,
+         default: false,
+         desc: 'Use control master for SSH connections',
+         aliases: '-c'
```
Comment (mjgiarlo) [inline bin/sdr:110]:
```suggestion
  option :skip_control_master,
         type: :boolean,
         default: false,
         desc: 'Skip checking for active SSH control master connection'
```

File: bin/sdr [hunk 2]
```diff
@@ -11,6 +11,21 @@ class CLI < Thor
     true
   end
 
+  no_commands do
+    def check_control_master(skip)
+      unless skip
```
Comment (mjgiarlo) [inline bin/sdr:16]: If the developer wants to skip this check, I wonder if the method should `return if skip`, just completely short-circuit the new method.

File: bin/sdr (hunk 1 above)
Comment (aaron-collier) [reply bin/sdr:110]: That makes sense to me. I'll update this and the readme.

File: bin/sdr (hunk 2 above)
Comment (aaron-collier) [reply bin/sdr:16]: I could be swayed either way, I was thinking this helps keep them from locking themselves out, but if they really want to... OK... LOL

File: bin/sdr [hunk 3]
```diff
@@ -11,6 +11,21 @@ class CLI < Thor
...
+        say("Checking SSH connections for #{options[:environment]} environment requires a control master.")
+        control_master = ask('Have you started a control master session: (y/n): ')
+        exit unless control_master.downcase == 'y'
+      end
+
+      return if `ps aux` =~ /ControlMaster=yes/
+
+      say('Control master not detected. Please start a control master session.')
+      exit
+    end
```
Comment (mjgiarlo) [inline bin/sdr:26]: I was thinking that an initial approach might be to use the `tty-prompt` library (already in `sdr-deploy`) to prompt the developer asking if they have a CM going already. If we're going to check for the CM ourselves, though, I'm not sure what the prompt gets us. Here's more or less what I had in mind, pseudocode-style:

```
Was skip flag provided?
* Yes. Do nothing.
* No.
  * Check for active CM. Is it present?
    * Yes. Do nothing.
    * No. Complain and abort.
```

Comment (aaron-collier) [reply bin/sdr]: Yeah, that works too. I was thinking the prompt just saves them rerunning the app, but that's a trivial savings for sure.

File: bin/sdr [hunk 4]
```diff
@@ -11,6 +11,21 @@ class CLI < Thor
...
 
+  no_commands do
+    def check_control_master(skip)
+      unless skip
+        say("Checking SSH connections for #{options[:environment]} environment requires a control master.")
+        control_master = ask('Have you started a control master session: (y/n): ')
+        exit unless control_master.downcase == 'y'
+      end
+
+      return if `ps aux` =~ /ControlMaster=yes/
```
Comment (mjgiarlo) [inline bin/sdr:17]: We should be able to use the `ssh` command to definitively check if the CM is present for a given named host (`sdr-infra`), and it ought to work in a cross-platform way. Like so:

```ruby
system('ssh -O check sdr-infra', err: File::NULL)
```

If ☝🏻 returns `true`, the CM is set up and running. If `false`, it's not.

Comment (mjgiarlo) [reply bin/sdr:17]: (And we think this is already working just fine given it's what we're using in dlss-capistrano: https://github.com/sul-dlss/dlss-capistrano/blob/cd4766654f600888ecb32bc3f18ff3735de9555e/lib/dlss/capistrano/tasks/control_master.rake#L23C8-L23C20)

File: bin/sdr (hunk 2 above)
Comment (mjgiarlo) [reply bin/sdr]: haha

File: README.md [hunk 5]
```diff
@@ -88,6 +88,7 @@ Options:
   -s, [--skip-update], [--no-skip-update]  # Skip update repos
   -e, --environment=ENVIRONMENT            # Environment (["qa", "prod", "stage"])
                                            # Possible values: qa, prod, stage
+      [--skip_control_master]              # Skip the control master check
```
Comment (mjgiarlo) [inline README.md:91]: This is what I'm seeing:

```suggestion
      [--skip-control-master]    # Skip checking for active SSH control master connection
```

File: README.md [hunk 6]
```diff
@@ -156,7 +157,7 @@ Options:
   -s, [--skip-update], [--no-skip-update]  # Skip update repos
   -e, --environment=ENVIRONMENT            # Deployment environment
                                            # Possible values: qa, prod, stage
-
+      [--skip_control_master]              # Skip the control master check
```
Comment (mjgiarlo) [inline README.md:160]: This is what I'm seeing:

```suggestion
      [--skip-control-master]    # Skip checking for active SSH control master connection
```

File: config/settings.yml [hunk 7]
```diff
@@ -1,3 +1,4 @@
+control_master_host: sdr-infra
```
Comment (mjgiarlo) [inline config/settings.yml:1]: I wonder if we should also toss a sentence or two at the README about this new CM stuff, including the config. But I'm approving either way. Thank you for working on this!

**Review by mjgiarlo — APPROVED** [review]

**Pull request merged**